src/Solve.cpp

```diff
@@ -345,31 +344,160 @@ class SolveExpression : public IRMutator {
 
         if (b_uses_var && !a_uses_var) {
             std::swap(a, b);
-        } else if (a_uses_var && b_uses_var) {
-            fail(T::make(a, b));
+            std::swap(a_uses_var, b_uses_var);
         }
 
-        if (a.same_as(op->a) && b.same_as(op->b)) {
-            expr = op;
+        const Add *add_a = a.as<Add>();
+        const Add *add_b = b.as<Add>();
+        const Sub *sub_a = a.as<Sub>();
+        const Sub *sub_b = b.as<Sub>();
+        const Mul *mul_a = a.as<Mul>();
+        const Mul *mul_b = b.as<Mul>();
+        const T *t_a = a.as<T>();
+        const T *t_b = b.as<T>();
+
+        expr = Expr();
+
+        if (a_uses_var && !b_uses_var) {
+            if (t_a) {
+                // op(op(f(x), a), b) -> op(f(x), op(a, b))
+                expr = mutate(T::make(t_a->a, T::make(t_a->b, b)));
+            }
+        } else if (a_uses_var && b_uses_var) {
+            if (equal(a, b)) {
+                // op(f(x), f(x)) -> f(x)
+                expr = a;
+            } else if (t_a) {
+                // op(op(f(x), a), g(x)) -> op(op(f(x), g(x)), a)
+                expr = mutate(T::make(T::make(t_a->a, b), t_a->b));
+            } else if (t_b) {
+                // op(f(x), op(g(x), a)) -> op(op(f(x), g(x)), a)
+                expr = mutate(T::make(T::make(a, t_b->a), t_b->b));
+            } else if (add_a && add_b && equal(add_a->a, add_b->a)) {
+                // op(f(x) + a, f(x) + b) -> f(x) + op(a, b)
+                expr = mutate(add_a->a + T::make(add_a->b, add_b->b));
+            } else if (add_a && add_b && equal(add_a->b, add_b->b)) {
+                // op(f(x) + a, g(x) + a) -> op(f(x), g(x)) + a;
+                expr = mutate(T::make(add_a->a, add_b->a)) + add_a->b;
+            } else if (add_a && equal(add_a->a, b)) {
+                // op(f(x) + a, f(x)) -> f(x) + op(a, 0)
+                expr = mutate(b + T::make(add_a->b, 0));
```

Comment:
I think the constant zero will only work for solving Int(32) expressions. Use make_zero(op->type) instead